vscode-dotnet-runtime-library/src/Utils/WebRequestWorker.ts

```diff
@@ -2,79 +2,140 @@
-            this.currentRequest.then((result) => {
-                if (result) {
-                    this.cachedData = result;
+            // Wrap the client with a retry interceptor. We don't need to return a new client, it should be applied automatically.
+            axiosRetry(uncachedAxiosClient, {
+                // Inject a custom retry delay to expoentially increase the time until we retry.
+                retryDelay(retryCount: number) {
+                    return Math.pow(2, retryCount); // Takes in the int as (ms) to delay.
                 }
-                this.currentRequest = undefined;
             });
-        }
-        return this.cachedData;
+
+            Debugging.log(`Axios client wrapped around axios-retry: ${uncachedAxiosClient}`);
+
+            this.client = setupCache(uncachedAxiosClient, {
+                storage: mementoStorage(extensionState),
+                ttl: cacheTimeToLive
+            });
+
+            Debugging.log(`Cached Axios Client Created: ${this.client}`);
     }
 
-    // Protected for ease of testing
-    protected async makeWebRequest(throwOnError: boolean): Promise<string | undefined> {
-        const options = {
-            url: this.url,
-            Connection: 'keep-alive',
-        };
+    public async getCachedData(retriesCount = 2): Promise<string | undefined> {
+        Debugging.log(`getCachedData() Invoked.`);
+        Debugging.log(`Cached value state: ${await this.isUrlCached()}`);
+        return await this.makeWebRequest(true, retriesCount);
+    }
 
-        try {
+    /**
+     * 
+     * @param cachedUrl 
+     * @returns true if the url was in the cache before this function executes, false elsewise.
+     * 
+     * @remarks Calling this WILL put the url data in the cache as we need to poke the cache to properly get the information.
+     * (Checking the storage cache state results in invalid results.)
+     * Returns false if the url is unavailable.
+     */
+    public async isUrlCached(cachedUrl : string = this.url) : Promise<boolean>
+    {
+        if(this.url === '')
+        {
+            return false;
+        }
+        try
+        {
+            const cachedState : boolean = (await this.client.get(cachedUrl, {timeout: 900})).cached; // 900 ms timeout is arbitrary but the expected worst case.
+            return cachedState;
+        }
+        catch (error) // The url was unavailable.
+        {
+            return false;
+        }
+    }
+
+    // Protected for ease of testing.
+    protected async makeWebRequest(throwOnError: boolean, retries: number): Promise<string | undefined> {
+        Debugging.log(`makeWebRequest Invoked. Requested URL: ${this.url}`);
+        try
+        {
             this.eventStream.post(new WebRequestSent(this.url));
-            const response = await request.get(options);
-            this.cacheResults(response);
-            return response;
-        } catch (error) {
+            const response = await this.client.get(
```

Comment:
It would be best to add some sort of promise.race or timeout besides the axios timeout as it doesn't cover all invalid url edge cases if this internal function is ever used incorrectly, but I tried this and it wasn't very successful, so probably not investing more time into that right now

Reply:
Added a timeout wrapper.